File: src/packages/control-ui/src/api/mobile_tunnel_supervisor.ts
export type MobileTunnelStatus =
  'DISCONNECTED' | 'CONNECTING' | 'CONNECTED' | 'RECONNECTING' | 'STOPPED';

export type SplitTunnelPolicy = 'ALLOW_ALL' | 'EXCLUDE_PACKAGES' | 'INCLUDE_PACKAGES';

export interface MobileSupervisorSettings {
  splitMode: SplitTunnelPolicy;
  packages: string[];
  primaryDns: string;
  fallbackDns: string;
  autoReconnect: boolean;
}

export class MobileTunnelSupervisor {
  state: MobileTunnelStatus = 'DISCONNECTED';
  private reconnectCount = 0;
  public settings: MobileSupervisorSettings;
  constructor(
    settings: MobileSupervisorSettings = {
      splitMode: 'ALLOW_ALL',
      packages: [],
      primaryDns: '1.1.1.1',
      fallbackDns: '8.8.8.8',
      autoReconnect: true,
    },
  ) {
    this.settings = settings;
  }

  startTunnel(): void {
    this.state = 'CONNECTING';
    this.state = 'CONNECTED';
    this.reconnectCount = 0;
  }

  stopTunnel(): void {
    this.state = 'STOPPED';
  }

  handleNetworkDrop(): void {
    if (this.settings.autoReconnect && this.state === 'CONNECTED') {
      this.state = 'RECONNECTING';
      this.reconnectCount++;
    } else {
      this.state = 'DISCONNECTED';
    }
  }

  isAppRouted(packageName: string): boolean {
    switch (this.settings.splitMode) {
      case 'ALLOW_ALL':
        return true;
      case 'EXCLUDE_PACKAGES':
        return !this.settings.packages.includes(packageName);
      case 'INCLUDE_PACKAGES':
        return this.settings.packages.includes(packageName);
    }
  }

  getEffectiveDns(useFallback: boolean): string {
    return useFallback ? this.settings.fallbackDns : this.settings.primaryDns;
  }

  getReconnectAttempts(): number {
    return this.reconnectCount;
  }
}

File: labs/mobile/donor-archive/tunnel/MobileTunnelSupervisor.kt
package com.luminet.android.tunnel

enum class AndroidTunnelState {
    DISCONNECTED, CONNECTING, CONNECTED, RECONNECTING, STOPPED
}

enum class SplitTunnelMode {
    ALLOW_ALL, EXCLUDE_PACKAGES, INCLUDE_PACKAGES
}

data class MobileSupervisorConfig(
    val splitMode: SplitTunnelMode = SplitTunnelMode.ALLOW_ALL,
    val packages: Set<String> = emptySet(),
    val primaryDns: String = "1.1.1.1",
    val fallbackDns: String = "8.8.8.8",
    val autoReconnect: Boolean = true
)

class MobileTunnelSupervisor(var config: MobileSupervisorConfig = MobileSupervisorConfig()) {
    var state: AndroidTunnelState = AndroidTunnelState.DISCONNECTED
        private set

    private var reconnectCount = 0

    fun startTunnel() {
        state = AndroidTunnelState.CONNECTING
        // Establish connection
        state = AndroidTunnelState.CONNECTED
        reconnectCount = 0
    }

    fun stopTunnel() {
        state = AndroidTunnelState.STOPPED
    }

    fun handleNetworkDrop() {
        if (config.autoReconnect && state == AndroidTunnelState.CONNECTED) {
            state = AndroidTunnelState.RECONNECTING
            reconnectCount++
        } else {
            state = AndroidTunnelState.DISCONNECTED
        }
    }

    fun isAppRouted(packageName: String): Boolean {
        return when (config.splitMode) {
            SplitTunnelMode.ALLOW_ALL -> true
            SplitTunnelMode.EXCLUDE_PACKAGES -> !config.packages.contains(packageName)
            SplitTunnelMode.INCLUDE_PACKAGES -> config.packages.contains(packageName)
        }
    }

    fun getEffectiveDns(useFallback: Boolean): String {
        return if (useFallback) config.fallbackDns else config.primaryDns
    }

    fun getReconnectAttempts(): Int = reconnectCount
}
